src/Tools/dotnet-gcdump/DotNetHeapDump/EventPipeDotNetHeapDumper.cs

```diff
@@ -0,0 +1,238 @@
+                        }
+                    };
+
+                    gcDumpSession.Source.Clr.GCStop += delegate (GCEndTraceData data)
+                    {
+                        if (data.ProcessID != processID)
+                        {
+                            return;
+                        }
+
+                        if (data.Count == gcNum)
+                        {
+                            log.WriteLine("{0,5:n1}s: DotNet GC Complete.", sw.Elapsed.TotalSeconds);
+                            dumpComplete = true;
+                        }
+                    };
+
+                    gcDumpSession.Source.Clr.GCBulkNode += delegate (GCBulkNodeTraceData data)
+                    {
+                        if (data.ProcessID != processID)
+                        {
+                            return;
+                        }
+
+                        etwDataPresent = true;
+
+                        if ((sw.Elapsed - lastEtwUpdate).TotalMilliseconds > 500)
+                        {
+                            log.WriteLine("{0,5:n1}s: Making  GC Heap Progress...", sw.Elapsed.TotalSeconds);
+                        }
+
+                        lastEtwUpdate = sw.Elapsed;
+                    };
+
+                    if (memoryGraph != null)
+                    {
+                        dumper.SetupCallbacks(memoryGraph, gcDumpSession.Source, processID.ToString());
+                    }
+
+                    listening = true;
+                    gcDumpSession.Source.Process();
+                    log.WriteLine("{0,5:n1}s: EventPipe Listener dying", sw.Elapsed.TotalSeconds);
+                });
+
+                // Wait for thread above to start listening (should be very fast)
+                while (!listening)
+                {
+                    readerTask.Wait(1);
+                }
+
+                for (; ; )
+                {
+                    if (readerTask.Wait(100))
+                    {
+                        break;
+                    }
+
+                    if (!etwDataPresent && sw.Elapsed.TotalSeconds > 5)      // Assume it started within 5 seconds.  
+                    {
+                        log.WriteLine("{0,5:n1}s: Assume no Dot Heap", sw.Elapsed.TotalSeconds);
+                        break;
+                    }
+                    if (sw.Elapsed.TotalSeconds > 20)       // Time out after 100 seconds. 
+                    {
+                        log.WriteLine("{0,5:n1}s: Timed out after 20 seconds", sw.Elapsed.TotalSeconds);
+                        break;
+                    }
+                    // TODO FIX NOW, time out faster if we seek to be stuck
+                    if (dumpComplete)
+                    {
+                        break;
+                    }
+                }
+
+                gcDumpSession.EndSession();
+
+                while (!readerTask.Wait(1000))
+                    Console.WriteLine("still reading...");
```

Comment:
I don't think we'd want this message spamming the console in a long wait.

Reply:
good catch!  Moved that to the verbose log.